### PR TITLE
Improve Digital adapter: custom creative renderer

### PR DESCRIPTION
**Pull request by jbartek25**
## Type of change
- [x] Feature

## Description of change
Improve Digital adapter: adding custom creative renderer

Doc PR: https://github.com/prebid/prebid.github.io/pull/3531

### CODE REVIEW

File: modules/improvedigitalBidAdapter.js [hunk 1]
```diff
@@ -10,6 +10,70 @@ const BIDDER_CODE = 'improvedigital';
 const RENDERER_URL = 'https://acdn.adnxs.com/video/outstream/ANOutstreamVideo.js';
 const VIDEO_TARGETING = ['skip', 'skipmin', 'skipafter'];
 
+class Razr {
```
Comment (FilipStamenkovic) [inline modules/improvedigitalBidAdapter.js:13]: Why creating a class with all `static` functions?  
You could just define and use functions without defining a class.

Comment (jbartek25) [reply modules/improvedigitalBidAdapter.js:13]: It was to group related methods and variables in a single entity. It looks nicer than for example prefixing the function and variable names with the "razr". Another options would be to replace the static class with a plain JS object. What is a preferred convention?

Comment (FilipStamenkovic) [reply modules/improvedigitalBidAdapter.js:13]: I don't think there is a preferred convention. You can leave like this. I just prefer clean functions over OOP (when OOP is not necessary).

Comment (jbartek25) [reply modules/improvedigitalBidAdapter.js:13]: I agree, using a static class is a bit of a stretch here. However, we have a brand new Improve Digital adapter that will replace the existing one hopefully in the next couple of weeks. I'll have the class replaced by an object literal similar to [here](https://github.com/prebid/Prebid.js/blob/master/modules/bluebillywigBidAdapter.js#L31) so if that's ok, please review as is.

Comment (jbartek25) [reply modules/improvedigitalBidAdapter.js:13]: @FilipStamenkovic can we proceed?

Comment (FilipStamenkovic) [reply modules/improvedigitalBidAdapter.js:13]: @ChrisHuie can you assign someone else to take a look at this PR? At the moment I don't have a spare time to dig into it.

Comment (jbartek25) [reply modules/improvedigitalBidAdapter.js:13]: Hey @ChrisHuie. Would it possible to get a rough ETA for when the PR will get reviewed? Thanks

Comment (ChrisHuie) [reply modules/improvedigitalBidAdapter.js:13]: @jbartek25 added a reviewer to take a look :)

Comment (Rothalack) [reply modules/improvedigitalBidAdapter.js:13]: I don't necessarily feel qualified to make the call on this question. From my perspective, it works and you're planning to replace this in the future, so I'm failing to see any consequence really.

File: modules/improvedigitalBidAdapter.js [hunk 2]
```diff
@@ -10,6 +10,70 @@ const BIDDER_CODE = 'improvedigital';
...
+    });
+
+    if (!this.loaded) {
+      const s = document.createElement('script');
+      s.type = 'text/javascript';
+      s.async = true;
+      s.src = 'https://razr.improvedigital.com/renderer.js';
+
+      const x = document.getElementsByTagName('script')[0];
+      x.parentNode.insertBefore(s, x);
```
Comment (patmmccann) [inline modules/improvedigitalBidAdapter.js:57]: please use ajax methods

**Comment by patmmccann** [comment]
One thing of note, and it doesnt disqualify this pr, but #6500 may already support your use case

File: modules/improvedigitalBidAdapter.js (hunk 2 above)
Comment (jbartek25) [reply modules/improvedigitalBidAdapter.js:57]: The script to be loaded will add logic to the page. Can you please elaborate on how and why this should be done with ajax?

**Comment by jbartek25** [comment]
@patmmccann  ah, I missed the remark about https://github.com/prebid/Prebid.js/pull/6500. Yes, we looked at it before and it suits our company needs better to have a custom renderer.

File: modules/improvedigitalBidAdapter.js (hunk 2 above)
Comment (patmmccann) [reply modules/improvedigitalBidAdapter.js:57]: because it is a prebid rule

Comment (patmmccann) [reply modules/improvedigitalBidAdapter.js:57]: you can use ajax like this https://github.com/prebid/Prebid.js/pull/7699/files or you can import the loadExternalScript function from utils. You cannot you the logic here.

Comment (jbartek25) [reply modules/improvedigitalBidAdapter.js:57]: Thank you. We'll review the options.

**Comment by patmmccann** [comment]
@jbartek25 here is another reference implementation

https://github.com/prebid/Prebid.js/blob/4378f2cc8b47b331e4b94a9fa3ec5f3962b38e71/modules/oneVideoBidAdapter.js#L139

File: modules/improvedigitalBidAdapter.js (hunk 2 above)
Comment (patmmccann) [reply modules/improvedigitalBidAdapter.js:57]: ty; I noted on the main thread, the logic in the oneVideo adapter would likely work for you as well.

Comment (jbartek25) [reply modules/improvedigitalBidAdapter.js]: code updated to rely on `executeRenderer`

**Comment by jbartek25** [comment]
@patmmccann does the code look good to you now?

**Review by patmmccann — APPROVED** [review]
yes, thank you!

**Pull request merged**